### ISSUE TITLE
Gravy - add automatic limiter option to avoid auto-resets

### ISSUE THREAD

**Issue by DaveBenham** [cosinekitty/sapphire#73]
High intensity input into Gravy with high resonance can result in some type of clipping, where the output suddenly goes to 0V for a short period of time, and then returns to normal - it really detracts from the otherwise wonderful output. It seems to happen when the cutoff frequency is tuned to match one of the components of the input, and it swells out of control, then slams shut (0V), then swells again, etc.

I first noticed this while experimenting with the Benjolin Oscillator pinging Gravy. The output is glorious. It may be my favorite VCF for the Benjolin if this can be fixed.

Maybe introduction of some tanh saturation clipping can help?

**Comment by cosinekitty** [comment]
It sounds like you are hitting Gravy's 100V limit where it automatically resets itself. It will drop to 0V output for a quarter second, then start working normally again (at least until another reset). I was concerned about the stability of the filter, especially with high resonance settings. Before I added auto-reset, there were cases where the filter would spiral out of control (NAN or INF) and get stuck there until I manually re-initialized it, which caused it to lose all its parameter settings too.

I am open to multiple modes, which could include a tanh (or bicubic) limiter, the existing auto-reset, or maybe something else too? Or maybe 100V is just too small a limit? (I thought 100V was rather large and would not be encountered by many people.)

In the meantime, maybe try a lower setting on the GAIN control? This might not be a satisfactory workaround, but it could help prevent the auto-resets until I get something new released.

**Comment by DaveBenham** [comment]
Yeah, I was just looking at the code and found the 100V limit. I agree, that should be more than enough.

The problem with gain adjustment is the inputs and cutoff frequency are constantly changing. Most of the time the output level is perfect, but then every now and then out of control resonance kicks in. I don't see a way to mitigate that other than to introduce a limiter within the module. A saturation limiter might actually be a nice feature, beyond solving this problem.

**Comment by cosinekitty** [comment]
I'm working on using the same automatic gain control (AGC) that I use for Tube Unit and Elastika. It is an adaptive limiter that should work well. @DaveBenham I'll post back here to let you know when I have a test build for you to play with.

**Comment by cosinekitty** [comment]
@DaveBenham I have a build for you to try. Gravy now has an automatic gain controller (AGC) in the right-click menu. It starts out disabled ("OFF") by default, for backward compatibility:

![image](https://github.com/user-attachments/assets/dafa134c-7952-4746-b3c9-2d9ae8ed5b74)

If you grab that slider and start pulling it to the left, you can adjust the automatic gain limiter to anywhere from 5V to 50V. Push it to the right of 50V and it will turn the limiter back off.

![image](https://github.com/user-attachments/assets/a57a6aed-1515-4121-926d-262270abbe7e)

Now if your output exceeds the limiter setting, the GAIN knob starts to glow red to let you know that the output level is being attenuated by the limiter:

![image](https://github.com/user-attachments/assets/53dd6ad9-edd9-4d59-af6c-d024145778a7)


The binaries are in the zip files at the bottom of [this artifact page](https://github.com/cosinekitty/sapphire/actions/runs/11470495228). As a reminder, you have to unzip the zip file, which inflates to one vcvplugin file. Then move that vcvplugin file to your Rack plugin folder as usual.

**Comment by DaveBenham** [comment]
Wonderful - so much better!

**Comment by cosinekitty** [comment]
Awesome! I made the limiter range much wider (5V - 50V) compared to the range I used for Elastika and Tube Unit (4V - 10V), because I find I sometimes need a rather hot signal to compensate for attenuation feeding into Galaxy reverb when its MIX is set to a high value.

Next I will go ahead and make the polyphonic version of Gravy. It should just fall into place with minor work, since all the hard stuff is already working in the existing Gravy module.
